txmgr,op-batcher: Improve logging & metrics

sebastianst opened the pull request:
- Standardize field names
- Add more fields to tx logging
- Add metrics for basefee and tip cap. Whenever we sample a basefee or suggested tip cap, we record these.
    - Note that the implementation was changed such that when waiting for confirmation, the chain head is always fetched, so that changes in the basefee can be recorded. This should help debugging transactions that are not getting included during basefee hikes.
- Log frame id and transaction hash together. This helps debugging transactions for a specific channel. 
    - In a follow-up, this will be further improved.
- Fix the `Queue` test to do assertions in the main testing goroutine, and remove a goroutine leak

Fixes https://github.com/ethereum-optimism/client-pod/issues/145

sebastianst committed:
txmgr: Remove goroutine use in queue test

Assertions should be made from the main test routine, if possible.
I had local test failures where assertions where made after the main
routine returned already.

Also, the existing test leaked goroutines for transactions that
weren't queued.

A waitgroup can be used here alternatively, but it's cleaner to just
assert in the main routine.

sebastianst commented on op-batcher/batcher/driver.go:
```diff
@@ -414,3 +412,19 @@ func (l *BatchSubmitter) l1Tip(ctx context.Context) (eth.L1BlockRef, error) {
+		case error:
+			fs = append(fs, "err", v)
+		default:
+			fs = append(fs, "ERROR", fmt.Sprintf("logFields: unknown type: %T", x))
```
Also works, I'm happy with either.

sebastianst replied:
you cannot use `l.Log`, this is a function, not a method on `l`. use `log` directly.

sebastianst replied:
I decided to keep logging this error to `ERROR` because it's then clearer what log call caused this.

sebastianst commented on op-service/txmgr/txmgr.go:
```diff
@@ -457,30 +463,32 @@ func (m *SimpleTxManager) queryReceipt(ctx context.Context, txHash common.Hash,
+	tip, err := m.backend.HeaderByNumber(ctx, nil)
 	if err != nil {
-		m.l.Error("Unable to fetch block number", "err", err)
+		m.l.Error("Unable to fetch tip", "err", err)
```
good point, adding

sebastianst commented on op-service/txmgr/metrics/tx_metrics.go:
```diff
@@ -141,6 +158,16 @@ func (t *TxMetrics) TxPublished(errString string) {
 	}
 }
 
+func (t *TxMetrics) RecordBasefee(basefee *big.Int) {
+	bff, _ := basefee.Float64()
+	t.basefee.Set(bff)
+}
+
+func (t *TxMetrics) RecordTipCap(tipcap *big.Int) {
+	tcf, _ := tipcap.Float64()
+	t.tipCap.Set(tcf)
+}
```
this returns the _accuracy_, not an error, and we don't need it here, so discard it.

trianglesphere commented on op-batcher/batcher/driver.go:
```diff
@@ -414,3 +412,19 @@ func (l *BatchSubmitter) l1Tip(ctx context.Context) (eth.L1BlockRef, error) {
+	for _, x := range xs {
+		switch v := x.(type) {
+		case txData:
+			fs = append(fs, "fid", v.ID(), "data_len", v.Len())
```
```suggestion
			fs = append(fs, "frame_id", v.ID(), "data_len", v.Len())
```
I'm assuming `fid` is frame ID here

sebastianst replied:
I meant to use `fid` for frame id, yeah. It's not used anywhere else in logging, so I just established `fid` for frame id. But I'm also happy with the more verbose field name `frame_id`, will change. In a follow-up, we can also streamline logging of channels as `channel_id`.

trianglesphere reviewed (approved):
I saw one confusing log field name, but the rest of the logging changes are big improvements. I also like the new L1 fee info metrics.

sebastianst committed:
txmgr,op-batcher: Improve logging & metrics

- Standardize field names
- Add more fields to tx logging
- Add metrics for basefee and tip cap
- Log frame id and transaction hash together
